# Pandas

In [1]:
# Importing numpy and pandas library
import numpy as np
import pandas as pd

### Pandas Series

One dimensional labeled array in Pandas series is capable of holding any type of data. Data alignment in Pandas is intrinsic. 

In [ ]:
# Understanding what goes into a pandas series
pd.Series?

In [ ]:
# Creating a pandas series from a list (similar to numpy.array)
my_list = [1,2,3]
print(my_list)
ds = pd.Series(my_list)
type(ds) # printing the type of pandas data series

In [ ]:
# Printing the output
ds

In [ ]:
# Syntax for creating a pandas series is pd.Series(data,index)

# Creating a pandas series from numpy array
my_array = np.linspace(0,10,11)
ds = pd.Series(my_array)
print(my_array)
print(80*'-')
ds # Pandas series prints the index and values (unlike numpy array)

In [ ]:
# Playing around with the index
my_array = np.linspace(0,10,11)
my_index1 = np.linspace(10,20,11)
my_index2 = np.linspace(100,110,11)
ds1 = pd.Series(data=my_array,index=my_index1) 
ds2 = pd.Series(data=my_array,index=my_index2) # Two series with same vlaues but different index

In [ ]:
ds1

In [ ]:
ds2

In [ ]:
# You can not compare the two dataframes (due to different indexes they are two different dataframes)
# For two pandas series to be comparable, they must be identical in all respects (index and values)
ds1 == ds2

In [ ]:
# Index can take any name (not only restricted to numbers)
my_array3 = np.array([1,2,3,4])
my_index3 = ["A B C D".split()]
ds3 = pd.Series(data=my_array3,index=my_index3)
ds3

In [ ]:
# Index can take any name (not only restricted to numbers)
my_array3 = np.array([1,2,3,4])
my_index3 = ["A B C D".split()]
ds3 = pd.Series(data=my_array3,index=my_index3)

my_array4 = np.array([1,2,3,4])
my_index4 = ["B C D E".split()]
ds4 = pd.Series(data=my_array4,index=my_index4)


In [ ]:
ds3

In [ ]:
ds4

In [ ]:
# When we try to add two series with different index values
ds3+ds4

In [ ]:
# What is NaN (arithmetic operations on nan)
np.nan 

In [ ]:
#Resetting index
ds.reset_index()
type(ds.reset_index())


### Pandas Data Frames

Recall that a numpy matrix was created by stacking numpy arrays. Similarly, a Pandas dataframe is constructed by stacking pandas series together.

In [ ]:
# Visualizing our data through a numpy matrix first
my_matrix = np.array([['Ronaldo', 2002, 'Brazil'],
                      ['Grosso', 2006, 'Italy'],
                      ['Iniesta', 2010, 'Spain'],
                      ['Gotze', 2014, 'Germany']])

my_matrix

In [ ]:
# Creating a pandas dataframe from scratch
df = pd.DataFrame([['Andy', 87, 'M'],
                   ['Bill', 67, 'M'],
                   ['Catherine', 93, 'F'],
                   ['David', 95, 'M'],
                   ['Emma', 50, 'F']],
                  columns=['Name', 'Score', 'Gender'])

df

In [ ]:
type(df)

In [ ]:
# Selecting a column
df['Score']

In [ ]:
# SQL way of selecting data (not the recommended way - not considered good a programming practice)
df.Score

In [ ]:
# Selecting two columns (spot the difference)
df[['Name','Score']]

In [ ]:
# Rearranging the column order
df = df[['Name','Gender','Score']] # not an inplace operation
df

In [ ]:
# Location operator - syntax is .loc[row,column]
df.loc[0,["Name","Score"]]

In [ ]:
# Modifying data using the Location operator
df.loc[0,"Score"] = 92
df

In [ ]:
# Modifying the index and using the loc operator
df.index = "A B C D E".split()
df
df.loc["A",:]
# Show error by changing A to 0

In [ ]:
# Now using the integer location operator - syntax .iloc[row_integer,column_integer]
df.iloc[0,:]

In [ ]:
# Showing column with iloc
df.iloc[:,1] # Rows and columns in iloc have to be integers (names will give an error)

In [ ]:
# When a series is compared to a constant (similar to numpy)
df[df['Score']>90]

### DataFrame Operations

In [ ]:
# Creating a new column
df = pd.DataFrame([['stock1',50,1000],
                   ['stock2',75,900],
                   ['stock3',125,400]],columns=['stock_name','price','shares'])
df
df['market_cap'] = df['price']*df['shares']

In [ ]:
# Sum function can be called on the series (and on the dataframe)
df['market_cap'].sum()

In [ ]:
# Create an empty dataframe column by broadcasting
df['pe_ratio'] = 0
df

In [ ]:
# Dropping the market_cap column
df.drop(columns=['market_cap']) # inplace=False is default
df

In [ ]:
# Dropping rows
df.drop(0) # not an inplace opearator

In [ ]:
# Sorting the dataframe on outstanding shares
df.sort_values(by=['shares'])

In [ ]:
# Setting index
df = df.set_index('stock_name')
df

In [ ]:
# Resetting index
df = df.reset_index()
df

In [ ]:
# Creating a dataframe 
df = pd.DataFrame(np.linspace(0,8,9).reshape(3,3),columns=['A B C'.split()])
df

In [ ]:
# Sum function 
df.sum(axis=0,skipna=True,min_count=2) 

### Data Analysis

In [ ]:
# Loading panel data through pandas
df = pd.read_csv("stock_data.csv",sep = ',')
df.shape

In [ ]:
# Viewing a sample of the dataframe (due to large size)
df.head()

In [ ]:
# Getting info on our dataframe
df.info()

In [ ]:
# Unique permnos
df['permno'].nunique()

In [ ]:
# Unique company names
df['company_name'].nunique()

In [ ]:
# Observe how the date is initially loaded 
df['date'][0]

In [ ]:
# Pandas has a convenient hack for dates
df['date'] = pd.to_datetime(df['date'])
df['date'][0]

In [ ]:
# Dataframe date format
df['date'].head()

In [ ]:
# Converting datetime from one format to another
df['date'].dt.strftime("%d-%m-%Y").tail()

In [ ]:
# writing dates in pandas
pd.Timestamp(year=2013,month=12,day=7)

In [ ]:
# Adding days to a date (or hours/minutes etc)
some_date = pd.Timestamp(year=2013,month=12,day=7)
print(some_date)
some_date + pd.Timedelta('3 days')

In [ ]:
# Creating a Series of dates (a very cool trick!)
my_date = pd.to_datetime(pd.Series(['13 July 2013','5 May 2010','15 March 2020']))
print(my_date)
#my_date + pd.tseries.offsets.MonthEnd(0)

In [ ]:
# Using Pandas date range to generate dates
pd.date_range(start='01 Jan 2021', end='31 Jan 2021')
#pd.date_range(start='01 Jan 2021', periods=18, freq= 'M')

In [ ]:
# Loading the dataframe again
df.tail()

In [ ]:
# Checking the dataframe types (from info)
df.info() # observe that prices and returns are still objects!

In [ ]:
# Converting price and returns to numbers
df['price'] = pd.to_numeric(df['price'],errors='coerce')
df['total_returns'] = pd.to_numeric(df['total_returns'],errors='coerce')

In [ ]:
# Checking info again
df.info()

In [ ]:
df.tail()

In [ ]:
# Rows that have NaNs
df.isnull() # Returns True in dataframe where values are NaN
# df.isnull().sum() gives count of NaN values

In [ ]:
# Dropping NaN values (involves dropping the whole row) - we drop the NaNs by writing command dropna()
df.dropna().tail() 

In [ ]:
# Method 1 for dealing with all NaN values
df.replace(np.nan,0).tail()

In [ ]:
# Method 2 for dealing with all NaN values
df['total_returns'] = df['total_returns'].fillna(0)
df['price'] = df['price'].fillna(method='ffill') # ffill is forward fill for handling NaN values
df.tail()

In [ ]:
# Filtering
df = df[df['permno']!=10001] # df and index start without 10001 permno

In [ ]:
# Exporting your dataframe to a csv
df.to_csv('out.csv')

### Merging

In [ ]:
# Creating a dataframe with stock names and stock price
df1 = pd.DataFrame([['A',100],
                    ['B',120],
                    ['C',50]],columns=['stock','price'])
df1

In [ ]:
# Creating a dataframe with stock name and stock PE ratio
df2 = pd.DataFrame([['B',10],
                    ['C',14],
                    ['D',1000]],columns=['stock','pe_ratio'])
df2

In [ ]:
# Simple merging syntax
pd.merge(left=df1,right=df2,on=['stock'],how='inner')
# inner - intersection
# outer - union
# left - keeps all left dataframe 
# right - keeps all right dataframe

### Groupby

Similar to the Pivot table in excel, this function lets you group/split data for calculations and transformations.

In [ ]:
# Reloading the dataset analyzed above (shortcut to parse the dates while reading the data)
df = pd.read_csv("stock_data.csv",sep = ',',parse_dates=['date'])

# Converting any letters into NaN
df['price'] = pd.to_numeric(df['price'],errors='coerce')
df['total_returns'] = pd.to_numeric(df['total_returns'],errors='coerce')

# Creating a new column
df['market cap'] = df['price']*df['share_outstanding']

# Printing the tail
df.tail(7)

In [ ]:
# Grouping by Company Name and counting number of data points available
df.groupby('permno')['total_returns']#.mean().sort_values()

In [ ]:
# Aggregating various statistics of returns for each PERMNO
df.groupby('permno')['total_returns'].agg([np.mean,np.median,np.std,np.var])

In [ ]:
# Grouping by dates and calculating average returns for each date
df.groupby('date')['total_returns'].mean()#.reset_index()

In [ ]:
# Binning market cap >200,000 USD
df.loc[df['market cap']<=200000,'category'] = 1
df.loc[df['market cap']>200000,'category'] = 2
df.head()

In [ ]:
# Grouping by two columns
df.groupby(['date','category'])['total_returns'].mean().head(15)#reset_index

### Stack/Unstack

In [ ]:
# Simple example of unstacking (from above example)
df.groupby(['date','category'])['total_returns'].mean().head(15)
# Execute the unstack command and show how 1 and 2 are now separate columns

In [53]:
# creating a new dataframe
df2 = pd.DataFrame(np.linspace(0,10,10).reshape(5,2),columns=['A','B'])
df2

A          B
0  0.000000   1.111111
1  2.222222   3.333333
2  4.444444   5.555556
3  6.666667   7.777778
4  8.888889  10.000000

In [54]:
# Stacking the columns 
df2.stack()

0  A     0.000000
   B     1.111111
1  A     2.222222
   B     3.333333
2  A     4.444444
   B     5.555556
3  A     6.666667
   B     7.777778
4  A     8.888889
   B    10.000000
dtype: float64

# Take Home Questions

- **Pandas Mul Function** - Use this dataframe created in the code chunk below. We have three columns with random float values. The goal is to multiply columns 1 and 2 with columns 3. Firstly, notice how running the function (df[['col1','col2']]*df['col3']) returns a matrix of NaN values. Now, use the .mul() function by passing df[['col1','col2']].mul(df['col3'],axis=0). Notice how we get the same NaN matrix when we change axis=1 instead. 

In [1]:
# Creating a dataframe for problem 1
import pandas as pd
import numpy as np
np.random.seed(123)
df = pd.DataFrame() # Creating a blank data-frame
df['col1'] = np.random.rand(10)
df['col2'] = np.random.rand(10)
df['col3'] = np.random.rand(10)
df

col1      col2      col3
0  0.696469  0.343178  0.634401
1  0.286139  0.729050  0.849432
2  0.226851  0.438572  0.724455
3  0.551315  0.059678  0.611024
4  0.719469  0.398044  0.722443
5  0.423106  0.737995  0.322959
6  0.980764  0.182492  0.361789
7  0.684830  0.175452  0.228263
8  0.480932  0.531551  0.293714
9  0.392118  0.531828  0.630976

In [8]:
df[['col1','col2']]*df['col3']
df[['col1','col2']].mul(df['col3'],axis=0)

col1      col2
0  0.441841  0.217712
1  0.243056  0.619278
2  0.164344  0.317726
3  0.336866  0.036465
4  0.519776  0.287564
5  0.136646  0.238342
6  0.354829  0.066023
7  0.156321  0.040049
8  0.141256  0.156124
9  0.247417  0.335571

- **Load CSV data** (*stock_data.csv*) - parse dates using the *pd.to_datetime* command, read the stock returns as numeric values using the *pd.to_numeric* command, drop all NaN values using the dropna(), and reset the index. We will be working with this dataframe (unless stated otherwise) for the remainder of the questions below.



In [2]:
df=pd.read_csv('stock_data.csv')

df['date'] = pd.to_datetime(df['date'])
df['price']=pd.to_numeric(df['price'],errors='coerce')
df=df.dropna()
df.reset_index(drop=True, inplace=True)
df.head()

permno company_name       date total_returns    price  share_outstanding
0   10001   ENERGY INC 2009-12-31      0.162621  10.3000               4361
1   10001   ENERGY INC 2010-01-31     -0.018932  10.0600               4361
2   10001   ENERGY INC 2010-02-28     -0.000656  10.0084               4361
3   10001   ENERGY INC 2010-03-31      0.020643  10.1700               4361
4   10001   ENERGY INC 2010-04-30      0.124385  11.3900               6070

- **Aggregating** - For each PERMNO, calculate the following statistics for the total_returns - sum, mean, std dev, and median. *Hint - use .groupby('permno') and then aggregate by the stats(.agg() function might be useful)*.



In [4]:
df['total_returns']=pd.to_numeric(df['total_returns'],errors='coerce')
df.groupby('permno')['total_returns'].agg([np.sum,np.mean,np.std,np.median])

sum      mean       std    median
permno                                        
10001   1.036046  0.011261  0.087456  0.005005
10006   1.838998  0.016274  0.088570  0.003731
10014   1.804638  0.066838  0.165212  0.071429
10028   4.170277  0.018371  0.199012  0.000000
10029   0.496859  0.015527  0.142866 -0.018682
10042   0.228737  0.001646  0.229307 -0.019048
10048   1.050106  0.013463  0.153554 -0.006173
10051   5.024822  0.016105  0.155655  0.005057
10057   4.203859  0.016294  0.124957 -0.004160
10064   0.277376  0.004334  0.150823 -0.005155
10066  -3.377112 -0.099327  0.251858 -0.112373
10071   2.403691  0.036420  0.117875  0.040795
10085   1.662973  0.019797  0.158935  0.006398
10092  -1.726790 -0.014759  0.162058 -0.030699
10097   0.196958  0.065653  0.204756 -0.024390
10102   1.922744  0.048069  0.098409  0.011778
10104   0.746993  0.009577  0.051786  0.008271
10108   1.703341  0.017381  0.100525  0.012995
10116   2.622359  0.018598  0.245405 -0.017241
10119   1.136931  0.005922  0.091410  0.003282
10120   1.547820  0.020638  0.112626  0.011278
10123   3.238436  0.021166  0.161788  0.000000
10125  -2.591487 -0.123404  0.326003 -0.153846
10126   0.224366  0.005904  0.082438  0.018688
10137   5.037922  0.011689  0.076735  0.013972

- **Filtering Text** - You can parse the company name column to search for companies that match your key words. A basic filter would be to check for "OIL", "GAS", and "ENERGY" companies. Write a program that returns the number of unique PERMNOs that match the above 3 key-words. *Hint*: Try using df['company_name'].str.contains(). If you are interested in text parsing in Python, **regex** (regular expression) is a powerful library you may want to try out!



In [5]:
df[df['company_name'].str.contains('|'.join(['OIL','GAS','ENERGY']))]['company_name'].unique()

array(['ENERGY INC', 'GAS NATURAL INC', 'ALLEGHENY ENERGY INC'],
      dtype=object)

- **Merging** - Create a new dataframe with the last available date and price for each PERMNO. For example, the only entry for PERMNO 10001 is 31-Jul-2017 (date) and 12.95 (price). Merge this new dataframe onto the original dataframe. Fill the missing/NaN values using pandas fillna function with a method of your choice.



In [6]:
last_df=df.sort_values('date').groupby('permno').tail(1)[['permno','date','price']]
df.append(last_df,ignore_index=True,sort=False).fillna(0)

permno          company_name       date  total_returns     price  \
0      10001            ENERGY INC 2009-12-31       0.162621  10.30000   
1      10001            ENERGY INC 2010-01-31      -0.018932  10.06000   
2      10001            ENERGY INC 2010-02-28      -0.000656  10.00840   
3      10001            ENERGY INC 2010-03-31       0.020643  10.17000   
4      10001            ENERGY INC 2010-04-30       0.124385  11.39000   
5      10001            ENERGY INC 2010-05-31       0.004829  11.40000   
6      10001            ENERGY INC 2010-06-30      -0.043421  10.86000   
7      10001       GAS NATURAL INC 2010-07-31       0.083333  11.72000   
8      10001       GAS NATURAL INC 2010-08-31      -0.111263  10.37100   
9      10001       GAS NATURAL INC 2010-09-30       0.076560  11.12000   
10     10001       GAS NATURAL INC 2010-10-31       0.032815  11.43990   
11     10001       GAS NATURAL INC 2010-11-30      -0.118436  10.04000   
12     10001       GAS NATURAL INC 2010-12-31       0.052291  10.52000   
13     10001       GAS NATURAL INC 2011-01-31       0.028992  10.78000   
14     10001       GAS NATURAL INC 2011-02-28       0.022727  10.98000   
15     10001       GAS NATURAL INC 2011-03-31       0.072404  11.73000   
16     10001       GAS NATURAL INC 2011-04-30      -0.038789  11.23000   
17     10001       GAS NATURAL INC 2011-05-31       0.028050  11.50000   
18     10001       GAS NATURAL INC 2011-06-30       0.008261  11.55000   
19     10001       GAS NATURAL INC 2011-07-31      -0.028139  11.18000   
20     10001       GAS NATURAL INC 2011-08-31      -0.013864  10.98000   
21     10001       GAS NATURAL INC 2011-09-30       0.005009  10.99000   
22     10001       GAS NATURAL INC 2011-10-31       0.005005  11.00000   
23     10001       GAS NATURAL INC 2011-11-30      -0.005000  10.90000   
24     10001       GAS NATURAL INC 2011-12-31       0.051835  11.42000   
25     10001       GAS NATURAL INC 2012-01-31      -0.019702  11.15000   
26     10001       GAS NATURAL INC 2012-02-29       0.005005  11.16080   
27     10001       GAS NATURAL INC 2012-03-31       0.048760  11.66000   
28     10001       GAS NATURAL INC 2012-04-30      -0.015000  11.44010   
29     10001       GAS NATURAL INC 2012-05-31      -0.084361  10.43000   
...      ...                   ...        ...            ...       ...   
2908   10137  ALLEGHENY ENERGY INC 2010-07-31       0.102514  22.80000   
2909   10137  ALLEGHENY ENERGY INC 2010-08-31      -0.010965  22.55000   
2910   10137  ALLEGHENY ENERGY INC 2010-09-30       0.094013  24.52000   
2911   10137  ALLEGHENY ENERGY INC 2010-10-31      -0.053834  23.20000   
2912   10137  ALLEGHENY ENERGY INC 2011-01-31       0.063531  25.78000   
2913   10014                     0 1977-03-31       0.000000   4.75000   
2914   10102                     0 1978-04-30       0.000000  49.87500   
2915   10006                     0 1984-05-31       0.000000  52.62500   
2916   10029                     0 1990-07-31       0.000000   6.12500   
2917   10125                     0 1991-08-31       0.000000   0.14063   
2918   10097                     0 1994-07-31       0.000000   6.50000   
2919   10048                     0 1994-12-31       0.000000   7.50000   
2920   10057                     0 1996-06-30       0.000000  30.00000   
2921   10064                     0 1997-02-28       0.000000  11.25000   
2922   10071                     0 1997-08-31       0.000000  84.68750   
2923   10120                     0 1998-04-30       0.000000  37.12500   
2924   10126                     0 2000-06-30       0.000000  31.25000   
2925   10116                     0 2000-08-31       0.000000   1.00000   
2926   10123                     0 2002-05-31       0.000000  18.80000   
2927   10066                     0 2002-07-31       0.000000   0.03000   
2928   10108                     0 2005-07-31       0.000000  35.89000   
2929   10085                     0 2006-03-31       0.000000  36.99000   
2930   101

- **Lagged Market Cap** - Create three new columns that contain the 6-month, 1-year, and 5-year lagged market cap. For example: PERMNO 10001 on 30-Jun-2016 will have the market cap as on 31-Dec-2015 (7.45 $\cdot$ 10505 = 78262.25) as its 6-month lag market cap. Perform a similar exercise to generate the 1-year and the 5-year lag market caps.



In [15]:
df['6-month']=df['price'].shift(6)*df['share_outstanding'].shift(6)
df['1-year']=df['price'].shift(12)*df['share_outstanding'].shift(12)
df['5-year']=df['price'].shift(60)*df['share_outstanding'].shift(60)
df.fillna(0)

permno          company_name       date  total_returns    price  \
0      10001            ENERGY INC 2009-12-31       0.162621  10.3000   
1      10001            ENERGY INC 2010-01-31      -0.018932  10.0600   
2      10001            ENERGY INC 2010-02-28      -0.000656  10.0084   
3      10001            ENERGY INC 2010-03-31       0.020643  10.1700   
4      10001            ENERGY INC 2010-04-30       0.124385  11.3900   
5      10001            ENERGY INC 2010-05-31       0.004829  11.4000   
6      10001            ENERGY INC 2010-06-30      -0.043421  10.8600   
7      10001       GAS NATURAL INC 2010-07-31       0.083333  11.7200   
8      10001       GAS NATURAL INC 2010-08-31      -0.111263  10.3710   
9      10001       GAS NATURAL INC 2010-09-30       0.076560  11.1200   
10     10001       GAS NATURAL INC 2010-10-31       0.032815  11.4399   
11     10001       GAS NATURAL INC 2010-11-30      -0.118436  10.0400   
12     10001       GAS NATURAL INC 2010-12-31       0.052291  10.5200   
13     10001       GAS NATURAL INC 2011-01-31       0.028992  10.7800   
14     10001       GAS NATURAL INC 2011-02-28       0.022727  10.9800   
15     10001       GAS NATURAL INC 2011-03-31       0.072404  11.7300   
16     10001       GAS NATURAL INC 2011-04-30      -0.038789  11.2300   
17     10001       GAS NATURAL INC 2011-05-31       0.028050  11.5000   
18     10001       GAS NATURAL INC 2011-06-30       0.008261  11.5500   
19     10001       GAS NATURAL INC 2011-07-31      -0.028139  11.1800   
20     10001       GAS NATURAL INC 2011-08-31      -0.013864  10.9800   
21     10001       GAS NATURAL INC 2011-09-30       0.005009  10.9900   
22     10001       GAS NATURAL INC 2011-10-31       0.005005  11.0000   
23     10001       GAS NATURAL INC 2011-11-30      -0.005000  10.9000   
24     10001       GAS NATURAL INC 2011-12-31       0.051835  11.4200   
25     10001       GAS NATURAL INC 2012-01-31      -0.019702  11.1500   
26     10001       GAS NATURAL INC 2012-02-29       0.005005  11.1608   
27     10001       GAS NATURAL INC 2012-03-31       0.048760  11.6600   
28     10001       GAS NATURAL INC 2012-04-30      -0.015000  11.4401   
29     10001       GAS NATURAL INC 2012-05-31      -0.084361  10.4300   
...      ...                   ...        ...            ...      ...   
2883   10137  ALLEGHENY ENERGY INC 2008-06-30      -0.082009  50.1100   
2884   10137  ALLEGHENY ENERGY INC 2008-07-31      -0.034125  48.4000   
2885   10137  ALLEGHENY ENERGY INC 2008-08-31      -0.063430  45.3300   
2886   10137  ALLEGHENY ENERGY INC 2008-09-30      -0.185528  36.7700   
2887   10137  ALLEGHENY ENERGY INC 2008-10-31      -0.180038  30.1500   
2888   10137  ALLEGHENY ENERGY INC 2008-11-30       0.169154  35.2500   
2889   10137  ALLEGHENY ENERGY INC 2008-12-31      -0.035177  33.8600   
2890   10137  ALLEGHENY ENERGY INC 2009-01-31      -0.018311  33.2400   
2891   10137  ALLEGHENY ENERGY INC 2009-02-28      -0.288809  23.6400   
2892   10137  ALLEGHENY ENERGY INC 2009-03-31      -0.013536  23.1700   
2893   10137  ALLEGHENY ENERGY INC 2009-04-30       0.118688  25.9200   
2894   10137  ALLEGHENY ENERGY INC 2009-05-31      -0.035494  25.0000   
2895   10137  ALLEGHENY ENERGY INC 2009-06-30       0.032000  25.6500   
2896   10137  ALLEGHENY ENERGY INC 2009-07-31      -0.017154  25.2100   
2897   10137  ALLEGHENY ENERGY INC 2009-08-31       0.047600  26.4100   
2898   10137  ALLEGHENY ENERGY INC 2009-09-30       0.009845  26.5200   
2899   10137  ALLEGHENY ENERGY INC 2009-10-31      -0.139517  22.8200   
2900   10137  ALLEGHENY ENERGY INC 2009-11-30      -0.036810  21.9800   
2901   10137  ALLEGHENY ENERGY INC 2009-12-31       0.075068  23.4800   
2902   10137  ALLEGHENY ENERGY INC 2010-01-31      -0.107751  20.9500   
2903   10137  ALLEGHENY ENERGY INC 2010-02-28       0.081146  22.6500   
2904   10137  ALLEGHENY ENERGY INC 2010-03-31       0.022075  23.0000   
2905   10137  ALLEGHENY ENERGY INC 2010-04-30      -0.053043  21

- **Resampling Frequency** - Convert the monthly dataframe that you loaded above into quarterly data and annual data. *Hint*: First, load and clean the data as per procedure mentioned above. Second, create a new column containing quarters corresponding to the given month (ex: 30-Nov-2010 is a quarter 4). Lastly, create a new dataframe "df_quarter" by aggregating data using the groupby command. For the quarterly dataframe your columns will be {permno, date, quarter, total_returns, market_cap}. Note that you will need to add the monthly returns to generate quarterly returns. Also note that you will need the quarter end market cap for each PERMNO each quarter. For example : In the Quarterly dataframe, PERMNO 10001 as of 31-Mar-2010 will have total returns 0.001055 (-0.018932-0.000656+0.020643) and the market cap 44351.37 (10.17$\cdot$4361). Perform the same procedure for creating data at an annual frequency.

In [81]:
df=pd.read_csv('stock_data.csv')
df['date'] = pd.to_datetime(df['date'])
df['price']=pd.to_numeric(df['price'],errors='coerce')
df['total_returns']=pd.to_numeric(df['total_returns'],errors='coerce')
df=df.dropna()
df.reset_index(drop=True, inplace=True)
df['year']=df.date.dt.year
df['quarter']=df.date.dt.quarter
df_quarter=df.groupby(['year','quarter','permno'])['total_returns'].agg(np.sum).reset_index()
temp=pd.merge(left=df,right=df_quarter,on=['permno','year','quarter'],how='outer')
quarter_end=df.groupby(['year','quarter','permno']).tail(1).reset_index()
quarter_end['market_cap']=quarter_end['price'].shift(0)*quarter_end['share_outstanding'].shift(0)
output=pd.merge(left=temp,right=quarter_end[['permno','year','quarter','market_cap']],on=['permno','year','quarter'],how='outer')
output[['permno', 'date', 'quarter', 'total_returns_y', 'market_cap']].rename(columns={'total_returns_y': 'total_returns'})

permno       date  quarter  total_returns  market_cap
0      10001 2009-12-31        4       0.162621    44918.30
1      10001 2010-01-31        1       0.001055    44351.37
2      10001 2010-02-28        1       0.001055    44351.37
3      10001 2010-03-31        1       0.001055    44351.37
4      10001 2010-04-30        2       0.085793    66028.80
5      10001 2010-05-31        2       0.085793    66028.80
6      10001 2010-06-30        2       0.085793    66028.80
7      10001 2010-07-31        3       0.048630    67531.76
8      10001 2010-08-31        3       0.048630    67531.76
9      10001 2010-09-30        3       0.048630    67531.76
10     10001 2010-10-31        4      -0.033330    82413.68
11     10001 2010-11-30        4      -0.033330    82413.68
12     10001 2010-12-31        4      -0.033330    82413.68
13     10001 2011-01-31        1       0.124123    91892.82
14     10001 2011-02-28        1       0.124123    91892.82
15     10001 2011-03-31        1       0.124123    91892.82
16     10001 2011-04-30        2      -0.002478    94144.05
17     10001 2011-05-31        2      -0.002478    94144.05
18     10001 2011-06-30        2      -0.002478    94144.05
19     10001 2011-07-31        3      -0.036994    89590.48
20     10001 2011-08-31        3      -0.036994    89590.48
21     10001 2011-09-30        3      -0.036994    89590.48
22     10001 2011-10-31        4       0.051840    93118.68
23     10001 2011-11-30        4       0.051840    93118.68
24     10001 2011-12-31        4       0.051840    93118.68
25     10001 2012-01-31        1       0.034063    95075.64
26     10001 2012-02-29        1       0.034063    95075.64
27     10001 2012-03-31        1       0.034063    95075.64
28     10001 2012-04-30        2      -0.126686    82375.60
29     10001 2012-05-31        2      -0.126686    82375.60
...      ...        ...      ...            ...         ...
2883   10137 2008-06-30        2       0.000996  8420033.41
2884   10137 2008-07-31        3      -0.283083  6209018.97
2885   10137 2008-08-31        3      -0.283083  6209018.97
2886   10137 2008-09-30        3      -0.283083  6209018.97
2887   10137 2008-10-31        4      -0.046061  5724168.44
2888   10137 2008-11-30        4      -0.046061  5724168.44
2889   10137 2008-12-31        4      -0.046061  5724168.44
2890   10137 2009-01-31        1      -0.320656  3924465.09
2891   10137 2009-02-28        1      -0.320656  3924465.09
2892   10137 2009-03-31        1      -0.320656  3924465.09
2893   10137 2009-04-30        2       0.115194  4345084.35
2894   10137 2009-05-31        2       0.115194  4345084.35
2895   10137 2009-06-30        2       0.115194  4345084.35
2896   10137 2009-07-31        3       0.040291  4494689.16
2897   10137 2009-08-31        3       0.040291  4494689.16
2898   10137 2009-09-30        3       0.040291  4494689.16
2899   10137 2009-10-31        4      -0.101259  3981339.24
2900   10137 2009-11-30        4      -0.101259  3981339.24
2901   10137 2009-12-31        4      -0.101259  3981339.24
2902   10137 2010-01-31        1      -0.004530  3900110.00
2903   10137 2010-02-28        1      -0.004530  3900110.00
2904   10137 2010-03-31        1      -0.004530  3900110.00
2905   10137 2010-04-30        2      -0.095565  3506893.72
2906   10137 2010-05-31        2      -0.095565  3506893.72
2907   10137 2010-06-30        2      -0.095565  3506893.72
2908   10137 2010-07-31        3       0.185562  4158959.80
2909   10137 2010-08-31        3       0.185562  4158959.80
2910   10137 2010-09-30        3       0.185562  4158959.80
2911   10137 2010-10-31        4      -0.053834  3935068.00
2912   10137 2011-01-31        1       0.063531  4381027.42

[2913 rows x 5 columns]